File: app/src/main/java/com/soho/sohoapp/feature/marketplaceview/components/MarketPlaceAdapter.java
package com.soho.sohoapp.feature.marketplaceview.components;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.soho.sohoapp.BaseViewHolder;
import com.soho.sohoapp.feature.home.BaseModel;

import java.util.List;

/**
 * Created by chowii on 14/8/17.
 */

public class MarketPlaceAdapter extends RecyclerView.Adapter<BaseViewHolder> {

    private final List<? extends BaseModel> propertyList;
    private final PropertyViewHolder.OnMarketplaceItemClickListener listener;

    public MarketPlaceAdapter(List<? extends BaseModel> propertyList, PropertyViewHolder.OnMarketplaceItemClickListener listener) {
        this.propertyList = propertyList;
        this.listener = listener;
    }

    @Override
    public int getItemCount() { return propertyList.size(); }

    @Override
    public int getItemViewType(int position) {
        return propertyList.get(position).getItemViewType();
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(
                parent.getContext())
                .inflate(viewType,
                        parent,
                        false);
        return new PropertyViewHolder(itemView, listener);
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, int position) {
        holder.onBindViewHolder(propertyList.get(holder.getAdapterPosition()));
    }

}

File: app/src/main/java/com/soho/sohoapp/feature/marketplaceview/components/MarketPlaceAdapter.kt
package com.soho.sohoapp.feature.marketplaceview.components

import android.support.v7.widget.RecyclerView.ViewHolder
import android.view.LayoutInflater
import android.view.ViewGroup

import com.soho.sohoapp.R
import com.soho.sohoapp.data.models.BasicProperty
import com.soho.sohoapp.feature.home.BaseModel
import com.soho.sohoapp.utils.PaginatedAdapter
import com.soho.sohoapp.utils.PaginatedAdapterListener

/**
 * Created by chowii on 14/8/17.
 */

class MarketPlaceAdapter(paginatedAdapterListener: PaginatedAdapterListener<BasicProperty>)
    : PaginatedAdapter<BasicProperty>(paginatedAdapterListener) {

    override fun viewHolderFormItem(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_property, parent, false)
        return PropertyViewHolder(itemView, paginatedAdapterListener)
    }

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        super.onBindViewHolder(holder, position)
        (holder as? PropertyViewHolder)?.onBindViewHolder(dataSet[position])
    }
}
